cmd/pebble/cmd_enter.go

```diff
@@ -0,0 +1,186 @@
+	Positional struct {
+		Cmd []string `positional-arg-name:"<subcommand>"`
+	} `positional-args:"yes"`
+}
+
+func init() {
+	optsHelp := map[string]string{
+		"run": "Start \"autostart\" services before executing subcommand",
+	}
+	for k, v := range sharedRunEnterOptsHelp {
+		optsHelp[k] = v
+	}
+	cmdInfo := addCommand("enter", shortEnterHelp, longEnterHelp, func() flags.Commander { return &cmdEnter{} }, optsHelp, nil)
+	cmdInfo.extra = func(cmd *flags.Command) {
+		cmd.PassAfterNonOption = true
+	}
+}
+
+func commandEnterFlags(commander flags.Commander) (enterFlags enterFlags, supported bool) {
+	supported = true
+	switch commander.(type) {
+	case *cmdExec:
+		enterFlags = enterFlagSilenceLogging
+	case *cmdHelp:
+		enterFlags = enterFlagNoServiceManager
+	case *cmdLs, *cmdPlan, *cmdServices, *cmdVersion:
+		enterFlags = enterFlagSilenceLogging | enterFlagProhibitServiceAutostart
+	case *cmdStart:
+		enterFlags = enterFlagKeepServiceManager
+	case *cmdStop:
+		enterFlags = enterFlagKeepServiceManager | enterFlagRequireServiceAutostart
+	default:
+		supported = false
+	}
+	return
+}
+
+func (cmd *cmdEnter) Execute(args []string) error {
+	if len(args) > 0 {
+		return ErrExtraArgs
+	}
+
+	runCmd := cmdRun{
+		sharedRunEnterOpts: cmd.sharedRunEnterOpts,
+	}
+	runCmd.setClient(cmd.client)
+
+	if len(cmd.Positional.Cmd) == 0 {
+		runCmd.run(nil)
+		return nil
+	}
+
+	runCmd.Hold = !cmd.Run
+
+	var (
+		commander flags.Commander
+		extraArgs []string
+	)
+
+	parser := Parser(cmd.client)
+	parser.CommandHandler = func(c flags.Commander, a []string) error {
+		commander = c
+		extraArgs = a
+		return nil
+	}
+
+	if _, err := parser.ParseArgs(cmd.Positional.Cmd); err != nil {
+		return err
+	}
+
+	enterFlags, supported := commandEnterFlags(commander)
+
+	if !supported {
+		return fmt.Errorf("enter: this command is not supported")
+	}
+
+	if enterFlags&enterFlagRequireServiceAutostart != 0 && !cmd.Run {
+		return fmt.Errorf("enter: this command can only run with autostarted services (hint: add --run to enter)")
+	}
+
+	if enterFlags&(enterFlagProhibitServiceAutostart|enterFlagNoServiceManager) != 0 && cmd.Run {
+		return fmt.Errorf("enter: this command cannot run with autostarted services (hint: remove --run from enter)")
+	}
+
+	if enterFlags&enterFlagNoServiceManager != 0 {
+		return commander.Execute(extraArgs)
+	}
+
+	if enterFlags&enterFlagSilenceLogging != 0 && !cmd.Verbose {
+		logger.SetLogger(logger.NullLogger)
+	}
+
+	runReadyCh := make(chan func(), 1)
+	runPanicCh := make(chan interface{})
+	var runStop func()
+	var runPanic interface{}
+
+	go func() {
+		defer func() { runPanicCh <- recover() }()
+		runCmd.run(runReadyCh)
+	}()
+
+	select {
+	case runStop = <-runReadyCh:
+	case runPanic = <-runPanicCh:
+		if runPanic == nil {
+			panic("internal error: pebble daemon stopped early")
+		}
+		panic(runPanic)
+	}
+
+	err := commander.Execute(extraArgs)
+
+	if err != nil || enterFlags&enterFlagKeepServiceManager == 0 {
+		runStop()
```

Comment:
Haven't reached the end of the code yet, so maybe this will be answered later, but how are we keeping the run command going if we're returning from `enter` early here?  In other words, the fact we're not asking the run command to stop explicitly won't do much for us if this command is about to return and exit from Pebble itself.

Reply:
This is handled by `runResultCh` (previously called `runPanicCh`) that's written to in the goroutine closure above after the `run` command ends, and that's being read from right after this statement.